File: model/src/main/java/io/github/vinicreis/model/enums/Result.java
package io.github.vinicreis.model.enums;

/**
 * Enum class that maps all possible results to be returned to clients.
 */
public enum Result {
    /**
     * Indicates that the request finished with no errors.
     */
    OK,
    /**
     * Indicates that some business or validation error occurred on the target.
     */
    ERROR,
    /**
     * Indicates that the client should make the request to other server to fetch an updated value from a key.
     */
    TRY_OTHER_SERVER_OR_LATER,
    /**
     * Indicates that the request key was not found on the server.
     */
    NOT_FOUND,
    /**
     * Indicates that an unexpected exception/error happened during the execution.
     */
    EXCEPTION
}

File: model/src/main/java/io/github/vinicreis/model/enums/Result.kt
package io.github.vinicreis.model.enums

sealed interface Result<T> {
    data class OkResult<T>(val data: T) : Result<T>
    data class ErrorResult<T>(val message: String) : Result<T>
    data class TryOtherServer<T>(val message: String, val timestamp: Long) : Result<T>
    data class NotFound<T>(val message: String) : Result<T>
    data class ExceptionResult<T>(val e: Exception) : Result<T>
}
